processing/src/main/java/io/druid/segment/IndexIO.java

```diff
@@ -374,8 +420,13 @@ public static void validateTwoSegments(File dir1, File dir2) throws IOException
         final Rowboat rb1 = it1.next();
         final Rowboat rb2 = it2.next();
         ++row;
+        if(rb1.getRowNum() != rb2.getRowNum()){
+          throw new IOException(String.format("Row number mismatch: [%d] vs [%d]", rb1.getRowNum(), rb2.getRowNum()));
+        }
         if(rb1.compareTo(rb2) != 0){
-          throw new IOException(String.format("Validation failure on row %d: [%s] vs [%s]", row, rb1, rb2));
+          if(!equalRowDims(rb1, adapter1, rb2, adapter2)){
```

Comment:
I think we should only compare dimension values here, and show the difference if there is a mismatch. Let's also add a comment in the code here to explain that this is for backwards compatibility with older indices that may result in differences with nulls when converting.